pymc4/distributions/state_functions.py

```diff
@@ -0,0 +1,250 @@
+            state_parts_st = tf.cast(state_parts_st, dtype=scales.dtype)
+            deltas = tf.unstack(
+                tf.map_fn(
+                    lambda x: generate_new_values(
+                        x[0], x[1]
+                    ),  # TODO: some issues with unpack and tf graph
+                    tf.stack([state_parts_st, scales], axis=1),
+                )
+            )
+            return tf.unstack(tf.cast(deltas, dtype=orig_dtype))
+
+    def __eq__(self, other) -> bool:
+        return self._name == other._name and self.scale == other.scale
+
+
+class GaussianRoundFn(Proposal):
+    """
+    Gaussian-Round proposal sub-class with the `_fn` that is sampling new proposal
+    from normal distribution N(0, 1) and rounding the values.
+
+    Parameters
+    ----------
+    scale: Union[List[Any], Any]
+        a `Tensor` or Python `list` of `Tensor`s of any shapes and `dtypes`
+        controlling the scale of the proposal distribution.
+    name: Optional[str]
+        Python `str` name prefixed to Ops created by this function.
+        Default value: 'categorical_uniform_fn'.
+    """
+
+    _name = "gaussian_round_fn"
+
+    def __init__(self, scale: Union[List[Any], Any] = 1.0, name: Optional[str] = None):
+        super().__init__(name)
+        self.scale = scale
+
+    def _fn(self, state_parts: List[tf.Tensor], seed: Optional[int]) -> List[tf.Tensor]:
+        scale = self.scale
+        with tf.name_scope(self._name or "gaussian_round_fn"):
+            scales = scale if mcmc_util.is_list_like(scale) else [scale]
+            if len(scales) == 1:
+                scales *= len(state_parts)
+            if len(state_parts) != len(scales):
+                raise ValueError("`scale` must broadcast with `state_parts`")
+
+            def generate_new_values(state_part, scale_part):
+                ndim = scale_part.get_shape().ndims
+                reduced_elem = tf.squeeze(
+                    tf.slice(
+                        scale_part, tf.zeros(ndim, dtype=tf.int32), tf.ones(ndim, dtype=tf.int32),
+                    )
+                )
+                delta = tfd.Normal(0.0, reduced_elem * 1.0).sample(
+                    seed=seed, sample_shape=(tf.shape(state_part))
+                )
+                state_part += delta
+                return tf.round(state_part)
+
+            state_parts_st = tf.stack(state_parts)
+
+            shape_ = state_parts_st.shape
+            inds_tile = tf.concat(
+                [tf.constant([shape_[0]]), tf.ones(shape_.ndims, dtype=tf.int32)], axis=0,
+            )
+            shape_ = (1,) + shape_[1:] + (shape_[0],)
+            scales = tf.tile(tf.broadcast_to(scales, shape_), inds_tile)[..., 0]
+
+            deltas = tf.unstack(
+                tf.map_fn(
+                    lambda x: generate_new_values(x[0], x[1]),
+                    tf.stack([state_parts_st, scales], axis=1),
+                )
+            )
+            return tf.unstack(deltas)
```

Comment:
The shape handling looks overly complicated here, how about something like:
```
def rounded_normal(state_part, scale):
    delta = tfd.Normal(0.0, tf.ones_like(state_part) * scale).sample(
                    seed=seed)
                )
    return tf.cast(delta, state_part.dtype) + state_part
new_state = tf.nest.map_structure(rounded_normal, state_parts, scales)
```

Reply:
yeap, It could be done easily with `tf.nest.map_structure`, fixing.

Reply:
Casting won't work because of the line here:
https://github.com/tensorflow/probability/blob/acb010fcc695bdfb74349777b94fa9e5f652375f/tensorflow_probability/python/distributions/poisson.py#L281.
It will crash for `x.dtype = tf.int32`